airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,257 @@
+        """
+        if isinstance(self._session, CachedLimiterSession):
+            self._session.cache.clear()  # type: ignore # cache.clear is not typed
+
+    # private method moved from HttpStream
+    def _dedupe_query_params(self, url: str, params: Mapping[str, str]) -> Mapping[str, str]:
+        """
+        Remove query parameters from params mapping if they are already encoded in the URL.
+        :param url: URL with
+        :param params:
+        :return:
+        """
+        if params is None:
+            params = {}
+        query_string = urllib.parse.urlparse(url).query
+        query_dict = {k: v[0] for k, v in urllib.parse.parse_qs(query_string).items()}
+
+        duplicate_keys_with_same_value = {k for k in query_dict.keys() if str(params.get(k)) == str(query_dict[k])}
+        return {k: v for k, v in params.items() if k not in duplicate_keys_with_same_value}
+
+    # private method moved from HttpStream
+    def _create_prepared_request(
+        self,
+        http_method: str,
+        url: str,
+        dedupe_query_params: bool = True,
+        headers: Optional[Mapping[str, str]] = None,
+        params: Optional[Mapping[str, str]] = None,
+        json: Optional[Mapping[str, Any]] = None,
+        data: Optional[Union[str, Mapping[str, Any]]] = None,
+    ):
+        if dedupe_query_params:
+            query_params = self._dedupe_query_params(url, params)
+        else:
+            query_params = params or {}
+        args = {"method": http_method, "url": url, "headers": headers, "params": query_params}
+        if http_method.upper() in BODY_REQUEST_METHODS:
+            if json and data:
+                raise RequestBodyException(
+                    "At the same time only one of the 'request_body_data' and 'request_body_json' functions can return data"
+                )
+            elif json:
+                args["json"] = json
+            elif data:
+                args["data"] = data
+        prepared_request: requests.PreparedRequest = self._session.prepare_request(requests.Request(**args))
+
+        return prepared_request
+
+    def _send_with_retry(self, request: requests.PreparedRequest, request_kwargs: Optional[Mapping[str, Any]] = None) -> requests.Response:
+        """
+        Backoff package has max_tries parameter that means total number of
+        tries before giving up, so if this number is 0 no calls expected to be done.
+        But for this class we call it max_REtries assuming there would be at
+        least one attempt and some retry attempts, to comply this logic we add
+        1 to expected retries attempts.
+        """
+        if self._backoff_strategy.max_retries is not None:
+            max_tries = max(0, self._backoff_strategy.max_retries) + 1
+
+        user_backoff_handler = user_defined_backoff_handler(max_tries=max_tries, max_time=self._backoff_strategy.max_time)(self._send)
+        backoff_handler = default_backoff_handler(
```

Comment:
Trying to integrate this with source-salesforce, the connector won't retry on HTTP status 420 because of [those lines](https://github.com/airbytehq/airbyte/blob/985c8396739bd84826e22e3359d945722c73afd8/airbyte-cdk/python/airbyte_cdk/sources/streams/http/rate_limiting.py#L43-L45) even though the ResponseAction is `ResponseAction.RETRY`

Reply:
Fixed by adding an httpclient-specific backoff handler